README.md
# airtable-clj

Clojure interface to [Airtable's API](https://airtable.com/api).

Work in progress!

## Usage

```clojure
(ns my-project.uses-airtable
  (:require [airtable-clj.core :as airtable]
            [environ.core :refer [env]))

(def api-key (:airtable-api-key env))
(def base (:airtable-base-id env))
(def table "My Table")

(def five-records
  (airtable/select {:api-key api-key
                    :base base
                    :table table
                    :max-records 5}))

(def just-one-record
  (airtable/retrieve {:api-key api-key
                      :base base
                      :table table
                      :record-id "rec123"}))

(airtable/create {:api-key api-key
                  :base base
                  :table table
                  :fields {"Foo" "boo"
                           "Bing" "bong"}
                  :typecast? false})
```

## License

Copyright © 2018 Evan Hahn

Distributed under the MIT License.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-03-28
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-03-28
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/airtable-clj/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/airtable-clj/compare/0.1.0...0.1.1


test/airtable_clj/core_test.clj
(ns airtable-clj.core-test
  (:require [clojure.test :refer :all]
            [environ.core :refer [env]]
            [airtable-clj.test-helpers :refer [mock-server take-mock-request]]
            [airtable-clj.core :as airtable]
            [airtable-clj.util :refer [handle-api-error]]))

(def ^:private fake-record-response
  {"id" "rec123"
   "fields" {"foo" "boo"
             "friend" "rec123"
             "bar" 123}
   "createdTime" "2018-04-20T16:20:00.000Z"})

(def ^:private fake-record
  {:id (fake-record-response "id")
   :fields (fake-record-response "fields")
   :created-time 1524241200000})

(def ^:private fake-404-response
  {:status 404
   :body {"error" "NOT_FOUND"}})

(def ^:private fake-record-404-response
  {:status 404
   :body {"error" {"type" "MODEL_ID_NOT_FOUND"
                   "message" "Not found!!"}}})

(def ^:private fake-delete-response
  {"id" "rec123"
   "deleted" true})

(def ^:private expected-user-agent (str "airtable-clj/" (env :airtable-clj-version)))

(deftest select-unit-test
  (testing "simple record selection"
    (let [responses [{:body {:records [fake-record-response]
                             :offset "rec456"}}]
          server (mock-server responses)
          result (airtable/select {:endpoint-url (:url server)
                                   :api-key "abc123"
                                   :base "base123"
                                   :table "My Table"})
          request (take-mock-request server)
          headers (:headers request)]
      (is (= "/v0/base123/My%20Table" (:uri request)))
      (is (= {} (:query-params request)))
      (is (= :get (:request-method request)))
      (is (= expected-user-agent (headers "user-agent")))
      (is (= "0.1.0" (headers "x-api-version")))
      (is (= "rec456" (:offset result)))
      (is (= (list fake-record) (:records result)))))
  (testing "selecting no records"
    (let [server (mock-server [{:body {:records []}}])
          result (airtable/select {:endpoint-url (:url server)
                                   :api-key "abc123"
                                   :base "base123"
                                   :table "My Table"})]
      (is (nil? (:offset result)))
      (is (empty? (:records result)))))
  (testing "adding additional options"
    (let [server (mock-server [{:body {:records []}}])
          _ (airtable/select {:endpoint-url (:url server)
                              :api-key "abc123"
                              :base "base123"
                              :table "My Table"
                              :fields ["Foo Boo" "Bing Bong"]
                              :filter-by-formula "NOT({X} = 2)"
                              :max-records 42
                              :page-size 69
                              ;; TODO
                              ;; :sort [{:field "Foo Boo"
                              ;;         :direction :desc}]
                              :view "My View"
                              :cell-format :json
                              :time-zone "Africa/Algiers"
                              :user-locale "bo"
                              :ignored-param "should not be included"})
          request (take-mock-request server)]
      (is (= "/v0/base123/My%20Table" (:uri request)))
      (is (= {"fields[]" ["Bing+Bong" "Foo+Boo"]
              "filterByFormula" ["NOT%28%7BX%7D+%3D+2%29"]
              "maxRecords" ["42"]
              "pageSize" ["69"]
              "view" ["My+View"]
              "cellFormat" ["json"]
              "timeZone" ["Africa%2FAlgiers"]
              "userLocale" ["bo"]} (:query-params request)))))
  (testing "calls out to handle-api-error"
    (let [server (mock-server [{:body {:records []}}])
          handle-api-error-arg (atom nil)]
      (with-redefs [handle-api-error #(reset! handle-api-error-arg %)]
        (airtable/select {:endpoint-url (:url server)
                          :api-key "abc123"
                          :base "base123"
                          :table "My Table"})
        (is (some? (:status @handle-api-error-arg)))))))

(deftest retrieve-unit-test
  (testing "retrieving a single record"
    (let [server (mock-server [{:body fake-record-response}])
          result (airtable/retrieve {:endpoint-url (:url server)
                                     :api-key "abc123"
                                     :base "base123"
                                     :table "My Table"
                                     :record-id "rec123"})
          request (take-mock-request server)
          headers (:headers request)]
      (is (= "/v0/base123/My%20Table/rec123" (:uri request)))
      (is (= {} (:query-params request)))
      (is (= :get (:request-method request)))
      (is (= expected-user-agent (headers "user-agent")))
      (is (= "0.1.0" (headers "x-api-version")))
      (is (= fake-record result))))
  (testing "getting a non-record 404 should throw"
    (let [server (mock-server [fake-404-response])]
      (is (thrown-with-msg? Throwable #"NOT_FOUND"
                            (airtable/retrieve {:endpoint-url (:url server)
                                                :api-key "abc123"
                                                :base "base123"
                                                :table "My Table"
                                                :record-id "rec123"})))))
  (testing "getting a record 404 should give nil by default"
    (let [server (mock-server [fake-record-404-response])]
      (is (nil? (airtable/retrieve {:endpoint-url (:url server)
                                    :api-key "abc123"
                                    :base "base123"
                                    :table "My Table"
                                    :record-id "rec123"})))))
  (testing "getting a record 404 can throw an error"
    (let [server (mock-server [fake-404-response])]
      (is (thrown-with-msg? Throwable #"NOT_FOUND"
                            (airtable/retrieve {:endpoint-url (:url server)
                                                :api-key "abc123"
                                                :base "base123"
                                                :table "My Table"
                                                :record-id "rec123"
                                                :throw-if-not-found? true})))))
  (testing "calls out to handle-api-error"
    (let [server (mock-server [{:body fake-record-response}])
          handle-api-error-arg (atom nil)]
      (with-redefs [handle-api-error #(reset! handle-api-error-arg %)]
        (airtable/retrieve {:endpoint-url (:url server)
                            :api-key "abc123"
                            :base "base123"
                            :table "My Table"
                            :record-id "rec123"})
        (is (some? (:status @handle-api-error-arg)))))))

(deftest create-unit-test
  (testing "record creation"
    (let [server (mock-server [{:body fake-record-response}])
          result (airtable/create {:endpoint-url (:url server)
                                   :api-key "abc123"
                                   :base "base123"
                                   :table "My Table"
                                   :fields {"Foo" "Boo"}})
          request (take-mock-request server)
          headers (:headers request)]
      (is (= "/v0/base123/My%20Table" (:uri request)))
      (is (= {} (:query-params request)))
      (is (= :post (:request-method request)))
      (is (= expected-user-agent (headers "user-agent")))
      (is (= "0.1.0" (headers "x-api-version")))
      (is (= "application/json" (headers "content-type")))
      (is (= {"fields" {"Foo" "Boo"}} (:body request)))
      (is (= fake-record result))))
  (testing "typecast? parameter set to true"
    (let [server (mock-server [{:body fake-record-response}])
          result (airtable/create {:endpoint-url (:url server)
                                   :api-key "abc123"
                                   :base "base123"
                                   :table "My Table"
                                   :fields {"Foo" "Boo"}
                                   :typecast? true})
          request-body (:body (take-mock-request server))]
      (is (= {"fields" {"Foo" "Boo"}
              "typecast" true} request-body))))
  (testing "typecast? parameter set to false"
    (let [server (mock-server [{:body fake-record-response}])
          result (airtable/create {:endpoint-url (:url server)
                                   :api-key "abc123"
                                   :base "base123"
                                   :table "My Table"
                                   :fields {"Foo" "Boo"}
                                   :typecast? false})
          request-body (:body (take-mock-request server))]
      (is (= {"fields" {"Foo" "Boo"}} request-body))))
  (testing "calls out to handle-api-error"
    (let [server (mock-server [{:body fake-record-response}])
          handle-api-error-arg (atom nil)]
      (with-redefs [handle-api-error #(reset! handle-api-error-arg %)]
        (airtable/select {:endpoint-url (:url server)
                          :api-key "abc123"
                          :base "base123"
                          :table "My Table"
                          :fields {"Foo" "Boo"}})
        (is (some? (:status @handle-api-error-arg)))))))

(deftest modify-unit-test
  (testing "modifying a record"
    (let [server (mock-server [{:body fake-record-response}])
          result (airtable/modify {:endpoint-url (:url server)
                                   :api-key "abc123"
                                   :base "base123"
                                   :table "My Table"
                                   :record-id "rec123"
                                   :fields {"Foo" "Boo"}})
          request (take-mock-request server)
          headers (:headers request)]
      (is (= "/v0/base123/My%20Table/rec123" (:uri request)))
      (is (= {} (:query-params request)))
      (is (= :patch (:request-method request)))
      (is (= expected-user-agent (headers "user-agent")))
      (is (= "0.1.0" (headers "x-api-version")))
      (is (= {"fields" {"Foo" "Boo"}} (:body request)))
      (is (= fake-record result))))
  (testing "clobbering a record"
    (let [server (mock-server [{:body fake-record-response}])
          _ (airtable/modify {:endpoint-url (:url server)
                              :api-key "abc123"
                              :base "base123"
                              :table "My Table"
                              :record-id "rec123"
                              :fields {"Foo" "Boo"}
                              :clobber? true})
          request (take-mock-request server)]
      (is (= :put (:request-method request)))
      (is (= {"fields" {"Foo" "Boo"}} (:body request)))))
  (testing "typecast? parameter set to true"
    (let [server (mock-server [{:body fake-record-response}])
          _ (airtable/modify {:endpoint-url (:url server)
                              :api-key "abc123"
                              :base "base123"
                              :table "My Table"
                              :record-id "rec123"
                              :fields {"Foo" "Boo"}
                              :typecast? true})
          request (take-mock-request server)]
      (is (= {"fields" {"Foo" "Boo"}
              "typecast" true} (:body request)))))
  (testing "typecast? parameter set to false"
    (let [server (mock-server [{:body fake-record-response}])
          _ (airtable/modify {:endpoint-url (:url server)
                              :api-key "abc123"
                              :base "base123"
                              :table "My Table"
                              :record-id "rec123"
                              :fields {"Foo" "Boo"}
                              :typecast? false})
          request (take-mock-request server)]
      (is (= {"fields" {"Foo" "Boo"}} (:body request)))))
  (testing "calls out to handle-api-error"
    (let [server (mock-server [{:body fake-record-response}])
          handle-api-error-arg (atom nil)]
      (with-redefs [handle-api-error #(reset! handle-api-error-arg %)]
        (airtable/modify {:endpoint-url (:url server)
                          :api-key "abc123"
                          :base "base123"
                          :table "My Table"
                          :record-id "rec123"
                          :fields {"Foo" "Boo"}})
        (is (some? (:status @handle-api-error-arg)))))))

(deftest delete-unit-test
  (testing "deleting a record"
    (let [server (mock-server [{:body fake-delete-response}])
          result (airtable/delete {:endpoint-url (:url server)
                                   :api-key "abc123"
                                   :base "base123"
                                   :table "My Table"
                                   :record-id "rec123"})
          request (take-mock-request server)
          headers (:headers request)]
      (is (= "/v0/base123/My%20Table/rec123" (:uri request)))
      (is (= {} (:query-params request)))
      (is (= :delete (:request-method request)))
      (is (= expected-user-agent (headers "user-agent")))
      (is (= "0.1.0" (headers "x-api-version")))
      (is (= {:id "rec123"} result))))
  (testing "getting a non-record 404 should throw"
    (let [server (mock-server [fake-404-response])]
      (is (thrown-with-msg? Throwable #"NOT_FOUND"
                            (airtable/delete {:endpoint-url (:url server)
                                              :api-key "abc123"
                                              :base "base123"
                                              :table "My Table"
                                              :record-id "rec123"})))))
  (testing "getting a record 404 should give nil by default"
    (let [server (mock-server [fake-record-404-response])]
      (nil? (airtable/delete {:endpoint-url (:url server)
                              :api-key "abc123"
                              :base "base123"
                              :table "My Table"
                              :record-id "rec123"}))))
  (testing "getting a record 404 can throw an error"
    (let [server (mock-server [fake-404-response])]
      (is (thrown-with-msg? Throwable #"NOT_FOUND"
                            (airtable/delete {:endpoint-url (:url server)
                                              :api-key "abc123"
                                              :base "base123"
                                              :table "My Table"
                                              :record-id "rec123"
                                              :throw-if-not-found? true})))))
  (testing "calls out to handle-api-error"
    (let [server (mock-server [{:body fake-delete-response}])
          handle-api-error-arg (atom nil)]
      (with-redefs [handle-api-error #(reset! handle-api-error-arg %)]
        (airtable/delete {:endpoint-url (:url server)
                          :api-key "abc123"
                          :base "base123"
                          :table "My Table"
                          :record-id "rec123"})
        (is (some? (:status @handle-api-error-arg)))))))


test/airtable_clj/util_test.clj
(ns airtable-clj.util-test
  (:require [clojure.test :refer :all]
            [cheshire.core :as json]
            [airtable-clj.util :refer [handle-api-error]]))

(defn- status [code] {:status code})

(defn- response [status body-map]
  {:status status
   :body (json/generate-string body-map)})

(deftest handle-api-error-test
  (testing "2xx status codes are fine"
    (handle-api-error (status 200))
    (handle-api-error (status 201)))
  (testing "401 errors"
    (is (thrown-with-msg? Throwable #"(?i)api key"
                          (handle-api-error (status 401)))))
  (testing "403 errors"
    (is (thrown-with-msg? Throwable #"(?i)forbidden"
                          (handle-api-error (status 403)))))
  (testing "404 errors"
    (is (thrown-with-msg? Throwable #"(?i)not found"
                          (handle-api-error (status 404))))
    (is (thrown-with-msg? Throwable #"bing bong"
                          (handle-api-error (response 404 {"error" "bing bong"}))))
    (is (thrown-with-msg? Throwable #"bing bong"
                          (handle-api-error (response 404 {"error" {"message" "bing bong"}})))))
  (testing "413 errors"
    (is (thrown-with-msg? Throwable #"(?i)too large"
                          (handle-api-error (status 413)))))
  (testing "422 errors"
    (is (thrown-with-msg? Throwable #"(?i)request cannot be processed"
                          (handle-api-error (status 422))))
    (is (thrown-with-msg? Throwable #"bing bong"
                          (handle-api-error (response 422 {"error" {"message" "bing bong"}})))))
  (testing "429 errors"
    (is (thrown-with-msg? Throwable #"(?i)too many requests"
                          (handle-api-error (status 429)))))
  (testing "500 errors"
    (is (thrown-with-msg? Throwable #"(?i)server error"
                          (handle-api-error (status 500))))
    (is (thrown-with-msg? Throwable #"(?i)server error"
                          (handle-api-error {:status 500
                                             :body "<html>Catastrophic error</html>"}))))
  (testing "503 errors"
    (is (thrown-with-msg? Throwable #"(?i)service unavailable"
                          (handle-api-error (status 503)))))
  (testing "other 4xx errors"
    (is (thrown-with-msg? Throwable #"(?i)bad request"
                          (handle-api-error (status 469)))))
  (testing "other 5xx errors"
    (is (thrown-with-msg? Throwable #"(?i)server error"
                          (handle-api-error (status 569))))))


test/airtable_clj/test_helpers.clj
(ns airtable-clj.test-helpers
  (:require [clojure.string :as string]
            [environ.core :refer [env]]
            [cheshire.core :as json])
  (:import [okhttp3.mockwebserver MockWebServer MockResponse]
           [java.net URL]
           [java.util.logging Logger Level]
           [java.util.concurrent TimeUnit]))

(def can-integration-test?
  (every? #(contains? env %) [:airtable-api-key
                              :airtable-base
                              :airtable-table
                              :airtable-record-id]))

(defn- disable-mockwebserver-logger []
  (let [logger-name (.getName MockWebServer)
        logger (Logger/getLogger logger-name)]
    (.setLevel logger Level/WARNING)))

(defn- ring-to-mock-response [{:keys [status body headers]}]
  (let [result (MockResponse.)
        headers (merge {"Content-Type" "application/json"} headers)]
    (if status (.setResponseCode result status))
    (doseq [[header-key header-value] headers]
      (.setHeader result header-key header-value))
    (.setBody result (json/generate-string body))
    result))

(defn- mock-request-to-map [mock-request]
  (if-not mock-request
    (throw (ex-info "No request to take from mock server" {})))
  (let [url (URL. "http" "example.com" -1 (.getPath mock-request))]
    {:uri (.getPath url)
     :path (.getFile url)
     :query-params (if-let [query (.getQuery url)]
                     (->> (string/split query #"&")
                          (map #(string/split % #"="))
                          (reduce (fn [params [k v]] (update params k conj v)) {})
                          (map (fn [[k v]] [k (sort v)]))
                          (into {}))
                     {})
     :request-method (-> (.getMethod mock-request) string/lower-case keyword)
     :headers (->> (.getHeaders mock-request)
                   .toMultimap
                   (map (fn [[k v]] [k (first v)]))
                   (into {}))
     :body (-> (.getBody mock-request) .readUtf8 json/parse-string)}))

(defn mock-server [responses]
  (disable-mockwebserver-logger)
  (let [server (MockWebServer.)
        url (str (.url server ""))
        mock-responses (map ring-to-mock-response responses)]
    (doseq [response mock-responses] (.enqueue server response))
    {:url url, :server server}))

(defn take-mock-request [{:keys [server]}]
  (mock-request-to-map (.takeRequest server 2 TimeUnit/SECONDS)))


src/airtable_clj/core.clj
(ns airtable-clj.core
  (:require [clj-http.client :as http]
            [cheshire.core :as json]
            [airtable-clj.util :refer [camelize-keyword
                                       parse-time
                                       make-url
                                       request-headers
                                       handle-api-error]]))

;; TODO: :as :json (strict?) for response coercion

(defn- format-record [record]
  {:id (record "id")
   :fields (record "fields")
   :created-time (parse-time (record "createdTime"))})

(defn- format-deletion [response]
  {:id (response "id")})

(defn- record-not-found? [response]
  (let [error-type (-> (:body response)
                       json/parse-string
                       (get-in ["error" "type"]))]
    (and (= 404 (:status response))
         (= "MODEL_ID_NOT_FOUND" error-type))))

(def ^:private select-options
  (->> [:fields
        :filter-by-formula
        :max-records
        :page-size
        :view
        :cell-format
        :time-zone
        :user-locale]
       (map (fn [param] [param (camelize-keyword param)]))
       (into {})))

(defn select
  "Select records from an base."
  [{:keys [api-key] :as options}]
  (let [url (make-url options)
        query-params (->> (select-keys options (keys select-options))
                          (map (fn [[k v]]
                                 [(k select-options)
                                  (if (keyword? v) (name v) v)]))
                          (into {}))
        http-options (cond-> {:headers (request-headers api-key)
                              :throw-exceptions false
                              :multi-param-style :array}
                       (seq query-params) (assoc :query-params query-params))
        response (http/get url http-options)
        _ (handle-api-error response)
        body (json/parse-string (:body response))]
    {:records (map format-record (body "records"))
     :offset (body "offset")}))

(defn retrieve
  "Retrieve a single record from a base."
  [{:keys [api-key record-id throw-if-not-found?] :as options}]
  (let [url (make-url options record-id)
        response (http/get url {:headers (request-headers api-key)
                                :throw-exceptions false})]
    (when-not (and (not throw-if-not-found?) (record-not-found? response))
      (handle-api-error response)
      (-> response :body json/parse-string format-record))))

(defn create
  "Create a record in a base."
  [{:keys [api-key fields typecast?] :as options}]
  (let [url (make-url options)
        body (cond-> {"fields" fields}
               typecast? (assoc "typecast" true))
        response (http/post url {:headers (request-headers api-key)
                                 :content-type :json
                                 :body (json/generate-string body)})]
    (handle-api-error response)
    (-> response :body json/parse-string format-record)))

(defn modify
  "Update a record in a base."
  [{:keys [api-key record-id fields clobber? typecast?] :as options}]
  (let [url (make-url options record-id)
        body (cond-> {"fields" fields}
               typecast? (assoc "typecast" true))
        http-fn (if clobber? http/put http/patch)
        response (http-fn url {:headers (request-headers api-key)
                               :content-type :json
                               :body (json/generate-string body)})]
    (handle-api-error response)
    (-> response :body json/parse-string format-record)))

(defn delete
  "Delete a record from a base."
  [{:keys [api-key record-id throw-if-not-found?] :as options}]
  (let [url (make-url options record-id)
        response (http/delete url {:headers (request-headers api-key)
                                   :throw-exceptions false})]
    (when-not (and (not throw-if-not-found?) (record-not-found? response))
      (handle-api-error response)
      (-> response :body json/parse-string format-deletion))))


src/airtable_clj/util.clj
(ns airtable-clj.util
  (:require [clojure.string :as string]
            [cheshire.core :as json]
            [environ.core :refer [env]])
  (:import [java.net URL]
           [java.text SimpleDateFormat]
           [com.fasterxml.jackson.core JsonParseException]))

(def api-version "0.1.0")
(def project-version (env :airtable-clj-version))
(def user-agent (str "airtable-clj/" project-version))

(defn request-headers [api-key]
  {"X-API-Version" api-version
   "Authorization" (str "Bearer " api-key)
   "User-Agent" user-agent})

(def ^:private date-format
  (SimpleDateFormat. "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"))
(defn parse-time [s]
  (-> (.parse date-format s) .getTime))

(defn make-url
  ([{:keys [endpoint-url base table]}]
   (let [base-url (URL. (or endpoint-url "https://api.airtable.com"))
         protocol (.getProtocol base-url)
         host (.getHost base-url)
         port (.getPort base-url)
         path (str "/v0/" base "/" table)]
     (-> (URL. protocol host port path) str)))
  ([options suffix]
   (str (make-url options) "/" suffix)))

(defn camelize-keyword [k]
  (-> (name k)
      (string/replace #"-(\w)" (fn ([[_ c]] (string/upper-case c))))))

(defn maybe-parse-json [s]
  (try
    (json/parse-string s)
    (catch JsonParseException _ nil)))

(defn handle-api-error [response]
  (let [status (:status response)
        body (maybe-parse-json (:body response))
        api-error (get body "error")
        api-error-message (if (string? api-error)
                            api-error
                            (get api-error "message"))
        error-message (cond
                        api-error-message api-error-message
                        (= status 401) "Unauthorized. Check your API key."
                        (= status 403) "Forbidden."
                        (= status 404) "Not found."
                        (= status 413) "Request body is too large."
                        (= status 422) "Request cannot be processed."
                        (= status 429) "Too many requests. Please try again later."
                        (= status 503) "Service unavailable."
                        (>= status 500) "Server error."
                        (>= status 400) "Bad request.")]
    (if error-message
      (throw (ex-info error-message {:type (get-in body ["error" "type"])
                                     :response response})))))
